Lps 123793

matuzalemsteles opened the pull request.

matuzalemsteles committed:
LPS-123793 Moves handler functions to utility files to be reused and exposes the useful methods from Fields and SettingsContext of LayoutProvider

matuzalemsteles committed:
LPS-123793 Moves the actions to the core folder and restructures the events with new nomenclature

matuzalemsteles committed:
LPS-123793 Add a new useConfig hook and ConfigProvider to store configuration data or properties without side effects

matuzalemsteles committed:
LPS-123793 Refactor the FormProvider and detach from the reducer to be configurable and add a context to expose the state

FormProvider is a key part of the project's foundation to bring configuration flexibility, this allows teams that use the Data Engine to compose their reducers and also use it as their Store allowing adding new use cases and evolve.

matuzalemsteles committed:
LPS-123793 Moves the components of the Data Engine foundation to the core folder

The idea is to have a core folder to create the separation/boundary of what is Data Engine and the form (not product) mechanism.

matuzalemsteles committed:
LPS-123793 Refactor the useCreateFieldChange hook to a thunk and remove the debounce

This hook was created because the debounce was keeping an old `pages` reference and this was causing problems, now we don't need a debounce anymore since we are migrating the top layer part and we will not have any performance problems.

Going back to thunk implementation to avoid creating hooks like this.

matuzalemsteles committed:
LPS-123793 Forces the value of readOnly in the Field component and avoids traversing all fields

The implementations in Metal.js and the old codebase abuse Visitor a lot in the hot path, it is a very expensive operation to apply, so we took advantage of the component in the hierarchy to do this and avoid using Visitor excessively.

Some uses of PagesVisitor in the hot path will be removed as we remove Metal.js and allocate these changes/aggregation of data in the right places.

matuzalemsteles committed:
LPS-123793 Simplifies the fieldClicked handler to avoid traversing pages to get field properties

matuzalemsteles committed:
LPS-123793 Moves the components, variants, and trunks to the respective Forms folder

The custom/forms folder is where the components and customizations that are exclusive to the Forms team are located. The same can be done for other teams or move to the respective module.

matuzalemsteles committed:
LPS-123793 Rename Renderer component to Page and consumes data directly from useFormState and useConfig

matuzalemsteles committed:
LPS-123793 Migrate events from LayoutProvider to the respective reducers

matuzalemsteles committed:
LPS-123793 Simplifies PageHeader implementation to handle localization and variation logic

Decouple the dependency of the EditablePageHeader component on the Page component and move the component to the MultiPagesVariant where it is only used.

Refactor the EditablePageHeader component to better handle localization, avoid using PageVisitor and use the index to change the title and description values according to the current language on the reducer.

Some of these changes are to simplify excess that deals with localization but also to remove the implementations that exist within the LayoutProvider that uses PageVisitor to change the title or description when the language changes.

The removal in LayoutProvider is not seen in this commit because these new changes will only reflect in Forms, the other products like Data Engine still use LayoutProvider and we will remove them in the next tickets as we migrate them.

matuzalemsteles committed:
LPS-123793 Adds the TranslationManager component to handle the component rendered via JSP and creates the reducer to handle languages

Moves the logic that changes the field values when the language changes to the reducer that was in LayoutProvider on the hot path in the getPages method, the idea is to keep the same resources but remove from the hot path in the new implementation of FormProvider in React.

The idea is to remove the implementations of StateSyncronizer and Main that dealt with listening for events from TranslationManager which is rendered via JSP, I moved this to a React component that will deal with listening for events and dispatch actions to FormProvider.

matuzalemsteles committed:
LPS-123793 Moves the implementation of the Title and description of the form to the React component

Removes the implementation of the Title and Description from the form in JSP and implements it in React, simplifies the orchestration that needed to be done to deal with the values of the fields and update when the language changes.

matuzalemsteles committed:
LPS-123793 Add the Management Toolbar component to control the component rendered via JSP

This is a step to move the components to React and also remove the main responsibilities of the Forms, it is still necessary to move some other methods that control some buttons of the component but it will be done in another ticket.

Once we have all this well aligned, we can also start to migrate the components rendered in the JSP to React to facilitate the integration and composition of the application.

matuzalemsteles committed:
LPS-123793 Adds the NavigationBar component to handle the component rendered via JSP

This is another move to remove the responsibilities of the Forms main component, it also simplifies the logic to deal with routing, we will be using the React Router and other components to take advantage of the React lifecycle to make it visible and not visible depending on the route.

This component only interacts with the component rendered via JSP, makes it simpler to remove it from the JSP and bring it to React in the near future.

matuzalemsteles committed:
LPS-123793 Replaces the use of usePage in favor of useFormState in the fields

Well, usePage now only contains relevant information from the current page and no longer the state of the application, the responsibility now lies with useFormState.

Some small changes in the organization of the declaration of variables in the ReactFieldBase to improve readability, it is still necessary to make many changes in this component that is taken care of a lot of responsibility.

matuzalemsteles committed:
LPS-123793 Refactor the Form component to compose with the new components and reducers

Well FormView is an application of the final result of the form visible to the user. It is also an application that handles reducers differently than Form Builder.

Not all reducers that are used in Forms are used in the User view, so we keep only the reducers that are used and also the proprietary reducers of the User view screen.

matuzalemsteles committed:
LPS-123793 Refactor the FormNoop component with the new base

FormNoop becomes LegacyFormBuilder which is just the component or compatibility layer to render the form renderer in a Metal.js component, this will be removed as we remove Metal.js from the codebase.

matuzalemsteles committed:
LPS-123793 Adds a configuration file to define the initial State and Config state

matuzalemsteles committed:
LPS-123793 Fixes error when changing the language back to recently typed text

Sidebar uses the focusedField information to use the settingsContext and render the fields, despite updating the data in the page structure in the builder, the focusedField was not updated and when it caused a rendering in the application it returned to the previous state that was in the focusedField.

The problem is that whenever you update the data in a field, you have to update the data in the focusedField and the page structure to avoid having old data problems, this can be a very silent problem, the focusedField must contain only the field id for reference but not all data, this requires more memory allocation and unnecessary duplication but for that we would have to change a lot of logic, leaving it as a suggestion for the team to change in the future.

matuzalemsteles committed:
LPS-123793 Adds the FormFieldSettings component

This component maintains a FormProvider to handle the structure of the Field's settingsContext, every Field contains the structure of a form that is used within the Sidebar to render a form where it allows editing the properties of the field rendered in the Builder.

FieldSettings needs its own Provider and reducers, it manipulates settingsContext page data and has different behaviors. But he also needs to talk to the root FormProvider and sync some data.

The use of ConfigProvider is still due to the fact that the Sidebar is in Metal and is not rendered on the same structure and cannot have visibility of the application's root Config.

Ideally, we need to have a more elegant way of keeping two FormProvider taking care of two different data but that are dependents.

Some reasons for having the sync is that the Builder root FormProvider reducers have the responsibility to update the information within the settingsContext which could be avoided and done within the FieldSettings reducers to avoid creating any state sync components. This is a research aria of the structural problem that needs to be studied and that could not be done during the migration from Metal.js to React

matuzalemsteles committed:
LPS-123793 Creates the Adapter component to use the Sidebar in Metal.js in React

This is a temporary Adapter that allows you to render the current Forms Sidebar in Metal.js in the React structure. Once we have the Data Engine Sidebar we can remove that implementation.

Some changes to the Sidebar in Metal.js were also made to replace the event names, also removing the Sidebar from Main.es.js that will be rendered in a React component. Present in the next commits.

matuzalemsteles committed:
LPS-123793 Adds the utility to parse the props from the root components to separate config from state

matuzalemsteles committed:
LPS-123793 Moves the FormsRuleBuilder component and refactor to use React Router and use the Form and Config providers

The FormsRulesBuilder was under the guard of the Data Engine but as I'm building the new Main component for Forms I'm moving near, this component is only being used in Forms and we need it near.

The same will be done for FormsRuleEditor and FormsRuleList in other commits when we move everything to the ddm-form-web module where it is the main place for the Forms application.

matuzalemsteles committed:
LPS-124147 Adds the Report component to control the visibility of the Entries page

The same as we did for ManagementToolbar and NavigationBar, removing the responsibility of main.es.js and taking advantage of the React and React Router lifecycle to deal with this. The idea in the future is that we can actually import the component into the React structure without using the taglib.

matuzalemsteles committed:
LPS-123797 Adds the FormBuilder component

This is the new FormBuilder component for Forms, it takes care of the Form page, centralizes the ManagementToolbar, TranslationManager, FormInfo, Pages and Sidebar components.

matuzalemsteles committed:
LPS-124147 Adds the new main component to Forms

Well, this is still not the final conclusion to completely remove main.es.js from form-web, it will still be removed in the next commits but it already includes the initial work using the new FormProvider, ConfigProvider and routing via React Router.

Due to React.js build conflicts plus Metal.js in the same module, we were unable to add this file to the form-web module where it should be currently, we will only move there when we remove the Metal.js components.

I also added the `initState` method which is in charge of transforming and adding some more data before initializing the application state, such as enabling successPages by default and adding a specific id for the page, rules like this were made in the main component and in the LayoutProvider in hot paths of the application, just bringing this to the beginning of the application to avoid problems of performance and mutability of the data since in LayoutProvider the data were different compared to the data that were in its child components due to these changes.

matuzalemsteles committed:
LPS-123793 Adjustments to form-renderer exports

matuzalemsteles committed:
LPS-124147 Fixes small blink adjustment for management bar when scrolling the page

matuzalemsteles committed:
LPS-124146 Rewrites AutoSave and StateSyncronizer in React.js

Well, this is a literal translation of the two components in React.js, few changes have been made to adapt to the React life cycle.

There are many DOM manipulations so that we can save the data in the DOM via `input[hidden]`, it may not be necessary anymore, a study in this area is needed to validate and remove these manipulations.

matuzalemsteles committed:
LPS-123793 Fixes error when saving the form

matuzalemsteles committed:
LPS-123793 Fixes error when not rendering the FormView

PaginationMode respects the value saved in Form Builder which is different from what should be viewed in FormView, we are only forcing the value of paginationMode for wizard, for now, there is no specification that this should be changed but if it is the backend must provide the name of the variation other than what is saved in Form Builder.

matuzalemsteles committed:
LPS-124147 Rewrites the implementation of controlling the Back button of the Control Menu

matuzalemsteles committed:
LPS-124147 Migrate the handlers from the management toolbar buttons in main.es.js to FormBuilder

Removes the handlers that existed in main.es.js for the respective component in React, still missing to move ShareFormModal to the respective Management Toolbar handler.

matuzalemsteles committed:
LPS-123793 SF

matuzalemsteles committed:
LPS-123793 Fixes field error is invisible in Form Builder when there are rules

When the Field has a rule to be visible when it satisfies some condition it is invisible in the Form Builder but this should only be applied to the User Preview, in this case, we are forcing all fields to be visible in the Form Builder.

matuzalemsteles committed:
LPS-123793 Fixes error in variations that accessing usePage information instead of useFormState and decouples Forms data from variations

Some of the Forms variations access some properties that are from Forms but that ended up being passed through Pages and Page so that it could get to the components, removing for instead of doing that accessing by useConfig or fetching data from useFormState depending on the property.

matuzalemsteles committed:
LPS-123793 Fixes the PAGE.CHANGE event payload

matuzalemsteles committed:
LPS-124147 Adds the implementation of ToastProvider to handle publication notifications for a form

The old implementation used an old ClayAlert v2 and used the wrong Lexicon pattern to display the success message for publishing the form, so applying the Lexicon pattern we use Toast with an action so that the user can open the form.

A ToastProvider was implemented instead of using the infra openToast because it has not yet implemented Toast with actions. We can remove this Provider when it is implemented.

matuzalemsteles committed:
LPS-124147 buildLang

matuzalemsteles committed:
LPS-124147 Fixes error when rendering the Form in the Element Set and saving

The Element Set page uses the Form to render with a different variation and does not pass information from autoSaveURL so we should not perform the autosave when that happens. Either way, it is not worth a lot of work to separate or disable Forms resources because the Element Set will be deprecated soon.

This commit also fixes a problem with the reference of rules that had in the `useEffect` that caused an autoSave at each rendering because the rules had a new array being instantiated at all times.

matuzalemsteles committed:
LPS-124147 Moves the default value implementation of the properties to the initialState

matuzalemsteles committed:
LPS-124147 Convert methods from DisplayContext to return JSON instead of String

Now that we are moving the application from Forms to React, we need the properties that were serialized to string to be just JSON so that we don't have to run JSON.parse at runtime to avoid deserialization bottlenecks.

matuzalemsteles committed:
LPS-124147 Instantiate the Forms application using the React Component taglib

Well, this actually deletes the main.es.js in Metal.js and the application adapters to use React only.

Use the Taglib provided by Frontend Infra to instantiate the application in React, this already deals with some use cases that we would have to deal with manually.

matuzalemsteles committed:
LPS-124147 Deletes Metal.js files and dependencies

Well, this consolidates the removal of Metal.js from the Forms application and now we can migrate the form-renderer components that are the responsibility of Forms to here.

matuzalemsteles committed:
LPS-124147 Forms responsibility components moved to the form-web module

Despite having many files being moved and edited, there are few changes the vast majority in the import and relocation of files to the respective modules.

The components that are exclusive to the Forms product are inside the form-web, components of `custom/forms` are components that may have become part of the Data Engine core in the future but the only customer is the product with the respective folder name or the component is too coupled to the product to be part of the core at the moment.

liferay-continuous-integration commented:
To conserve resources, the PR Tester does not automatically run for every pull.

If your code changes were already tested in another pull, reference that pull in this pull so the test results can be analyzed.

If your pull was never tested, comment &quot;ci:test&quot; to run the PR Tester for this pull.

matuzalemsteles committed:
LPS-124147 Change the name of the file that links to FormView

This file had a nomenclature of `index` which can cause a lot of confusion in the codebase passing on the idea that it is the main file that exports or initializes the application a very common standard in JS codebase, since that is not the purpose, I'm changing to the file that lives up to its main idea.

matuzalemsteles committed:
LPS-123793 Simplifies the way to customize the actions of the field and added the confirmation thunk to delete a field

This commit implements two changes, `fieldActions` is no longer exposed through the FormProvider state and adds the `fieldDelete` thunk.

- `fieldActions` is configured via props using `<Pages />`, this makes it easier when the action needs information from the FormProvider itself and access to other properties of the application.
- `fieldDelete` thunk to check if the field being deleted exists within a rule to show the confirmation message before dispatching.

matuzalemsteles committed:
LPS-123793 Removes unused events from LayoutProvider

These events are used only in Forms and we no longer need them in LayoutProvider, removing them since it becomes stale and was implemented in React.

See the history of commits with LPS-123793.

matuzalemsteles committed:
LPS-123793 Fixes rebase errors

matuzalemsteles committed:
LPS-123793 Fixes error when rendering all FormView in wizard mode

This creates a mapping between pagination in builder and view mode, in Forms the `multi_pages` is only in Builder but in View it is represented with a wizard.

diegonvs committed:
LPS-124149 Migrate ShareFormModal to React

We can have two possible ways to instantiate a ShareFormModal:

* Using openShareFormModal function
* Using directly ShareFormModalBody/Modal buttons React component

The first one was necessary for the use case of view_form_instance.jsp file and the second one is used for calling the ModalContext redux with correct values.

We're reusing the same ModalContext for avoiding have multiple ModalContexts around the application and it prevent duplicate errors

matuzalemsteles committed:
LPS-124147 Refactor FormsMultiPanelSidebar and fixes errors from Sidebar events

This brings the MultiPanelSidebar that was in main.es.js to Forms with flag but I made some changes to not have a component with its own context and adapter, it just simplifies the use of the component to remove own implementations on the Forms side.

I moved the components referring to MultiPanelSidebar which is the plugin to ddm-form-web for the responsibility of the Forms which is a palliative until the Forms can consume the Taglib of the Data Engine or something related.

It also fixes some problems of importing or exporting methods and files that were lost during rebase. A change in the mapping of events that is for compatibility with the LayoutProvider.

matuzalemsteles committed:
LPS-124149 Renames the folder name in favor of the default in the codebase

This is just a change of the folder name, instead of using CamelCase for the folder with the components use `like-this` which is something more common throughout the Liferay code base.

matuzalemsteles committed:
LPS-124147 Apply source format

matuzalemsteles committed:
LPS-124147 Persist the rule in progress when browsing pages

Well, this change centralizes which rule is being edited or created in the FormProvider state, this to be able to redirect the user to the rule's edit/creation page when he returns from another page navigated by the React Router.

matuzalemsteles committed:
LPS-124147 Migrates the FormRuleList and FormRuleEditor for the responsibility of the Form module

Well, the idea is to migrate the two components for the responsibility of Forms, we had added it in the data-engine-taglib module because the forms-web module was unable to add components with React since we can now add it after we remove Metal.js.

matuzalemsteles committed:
LPS-123793 Revert the removed getRules() method

Only by reversing this change, apparently getRules() is also used in the Data Engine.

matuzalemsteles committed:
LPS-124147 Added the ID to the Title and Description to adjust the test selector

matuzalemsteles committed:
LPS-123793 Fixes form-renderer module tests

The test of the PageHeader of MultiPagesVariant had to undergo some small changes to adapt to the new use, the good side is that we were able to test the corresponding reducer as well.

diegonvs committed:
LPS-124149 Improves accessibility when the Copy button uses an check icon as a value

diegonvs committed:
LPS-124149 Fixes errors on Email multi-select

It fixes:

* Now We're able to remove a single email on the multi-select
* Before, the users was unable to use the multi-select when any data was retrieved from autocompleteUsers service but now even though the service fails the user is able to send the email for the given email addresses

diegonvs committed:
LPS-124149 Adds tests for ShareFormModal

matuzalemsteles committed:
LPS-124149 Moves the ShareFormModal test file to the correct test folder

matuzalemsteles committed:
LPS-124147 Removes the use of JSP-controlled activeNavItem

The best case for this is to control via JSP to have the rendering on the server but the way this is done to get the active item nav from request is having problems with Senna.js, when a navigation is done the JavaScript has to update the query of the URL but in some use cases Senna.js checks the previous URL and the new URL after browsing and this causes a refresh of the page.

- Go to Form List
- Create new Form
- Add new field
- Navigate to Rules (Refresh happens)

That's because Senna.js listens to `onPopState` and now we're using the React Router that also handles history.

matuzalemsteles committed:
LPS-124147 Fixes validation failure event

The reducer was waiting for a PAGE.VALIDATION_FAILED event but formValidate was issuing the PAGE_VALIDATION_FAILED event.

This also moves the reducer to the core.

matuzalemsteles committed:
LPS-123793 Preserves the editable false value in the FieldSet

The editable has to be preserved when it is set to false in the root component, the FieldSet conditionals are only valid when the editable is true, that is, in edit mode.

matuzalemsteles committed:
LPS-123793 Fixes the types of Panel and FieldBase actions

matuzalemsteles committed:
LPS-124147 Fixes the initial value of availableLanguageIds

The value of `themeDisplay.getLanguageId()` is the one used by the user and in Forms it is considered the default languageId of the website.

matuzalemsteles committed:
LPS-123793 Fixes the action of DUPLICATE the field when fired by the Sidebar

The Sidebar in Metal.js, which is temporary, has no visibility of the active page, so instead of taking the activePage from payload, get the pageIndex of the Field which is just used to add the copy of the field below the original.

matuzalemsteles committed:
LPS-124147 Fixes error of initialState or initialConfig not overwriting properties with undefined values

By using destructuring to parse the props, by default it will add the property with an undefined value if it is not passed, so we clear properties with these values to allow the initialState to enter the action.

matuzalemsteles committed:
LPS-124147 Removes the JSP Form Info from the element set

This is no longer necessary because it was migrated to React.

matuzalemsteles committed:
LPS-123793 Computes changing localizedValue only when settingsContext does not exist

The language reducer is used for the main application, FormView, Forms app... and the Sidebar itself but the action of changing the language is different for the Builder compared to the other cases, we do not need to compute the change of localizedValue when it is executed in the root FormProvider because the properties do not exist but will be computed for the other cases.

matuzalemsteles committed:
LPS-123793 Fixes the computation of the action of adding Element Set that was not a pure function

matuzalemsteles committed:
LPS-123793 Allows to change the defaultLanguageId at runtime

After rebasing, the ticket LPS-128585 added the master to allow changing the defaultLanguageId, it was necessary to move the defaultLanguageId to the state instead of the config since it can be changed at runtime so as not to cause regression.

matuzalemsteles committed:
LPS-123793 Fixes error when creating field group

It was missing properties to create the field group, the fieldTypes was always empty because it gets from the state instead of the config.

diegonvs committed:
LPS-123793 Traverse missing `published` argument on useAutoSave and fixes a small typo on `updateAutoSaveMessage` function

Without this change, auto save message will not be correct showing always saved without the draft

matuzalemsteles committed:
LPS-123793 Fixes the selectors of Delete the field with the Confirmation Modal

This particular modal underwent markup changes to use the most current Modal markup.

matuzalemsteles committed:
LPS-124149 Fixes the size of the link input

matuzalemsteles committed:
LPS-123793 Fixes the unpublishForm selector and macro

matuzalemsteles committed:
LPS-123793 Fixes error when synchronizing the language state in the Sidebar

Well, since the defaultLanguageId became part of the state, we also need to pass the defaultLanguageId to the state of the Sidebar so that it does not update with an undefined value.

matuzalemsteles committed:
LPS-123793 Fixes error when changing field type

Well, this change separates the result of an evaluation in two reducers, previously this was the `focusedFieldEvaluationEnded` event that dealt with:

- Evaluation of a field
- Change field type from Sidebar

This change separates the responsibility to reduce the redundancy that existed in computing the data mainly for changing the type of the field that the payload already came with the values ready.

diegonvs committed:
LPS-123793 Move DragLayer component to FormBuilder component

Without this change the DragPreview component will not be rendered correctly. Some classes will not work properly due to DOM order

diegonvs committed:
LPS-123793 Pass the correct pages value for adding a field when drag/drop around the builder

diegonvs committed:
LPS-123793 Defines the initial value of `allowNestedFields` property

This property comes defined on data-engine based applications but it's not the same for Forms

diegonvs committed:
LPS-123793 Typo  EVENT_TYPES.PAGE.VALIDATION_FAILED -> CORE_EVENT_TYPES.PAGE.VALIDATION_FAILED

matuzalemsteles committed:
LPS-123793 Fixes error when adding options from some field

Well, the options fields have an inconsistency with the localization system, the value property does the same thing as the localizedValue property.

When a field change occurs for an Option field, it will come with the value `value: {[languageId]: '...'}` but we need to remove that languageId from the value. Because the corresponding field does not expect this, for example the `columns` and `rows` of the Grid field.

matuzalemsteles committed:
LPS-124147 Apply source format

matuzalemsteles committed:
LPS-124147 Add page normalization to remove uniqueness from the data that comes from the backend

Well, this is more of a stopgap the data structure of the Forms backend is returning a `JSONArray`, in the backend the structure is a String that is converted to JSONArray but this should perhaps be a proper structure to handle the data better, using Map, List... this can probably be removed soon.

matuzalemsteles committed:
LPS-124147 Normalize the options field in the page structure before starting the application

Well, the Options Field is very complex and needs a lot of improvements to deal with localization and state control, it has a lot of edge cases that made it necessary to add conditionals in the rest of the application to control the state or persist behaviors.

This is one of them that was on LayoutProvider and we need to keep it here so we don't have to refactor the Options field now because we don't have time. We may need to move this fix to another place within the core, because the Data engine will need the same fix, keeping it here while the Data Engine still uses the components in Metal.js.

matuzalemsteles committed:
LPS-124147 Synchronizes the rule state of SettingsContext

FormFieldSettings needs to synchronize the rules state of the Field's SettingsContext when a change field type occurs.

matuzalemsteles committed:
LPS-124149 Waits for the element to be present before getting the value in the getPublishedFormPublicURL macro

This is an attempt to remove the flaky tests in the CI that are always failing to get the input value in Modal, the assumption is that when it gets the value, the element has not finished rendering yet.

Modal has two body rendering phases, first, it creates the opening animation and after the animation is finished it will render the body.

matuzalemsteles committed:
LPS-124147 Ignore value when Change Field Type

We should not consider the old field's value, because they may not match the current field for the new field.

The value can be string, array, and other type.

matuzalemsteles committed:
LPS-124149 Use Form.getPublishedFormPublicURL instead of using the selector directly

The macro is already handling this Selector better, we have to wait for the element to appear to get its value.
